internal/lib/checkpoint.go

```diff
@@ -0,0 +1,233 @@
+}
+
+// getDiff returns the file system differences
+// Copied from libpod/diff.go and simplified for the checkpoint use case
+func (c *ContainerServer) getDiff(id string) ([]archive.Change, error) {
+	layerID, err := c.GetContainerTopLayerID(id)
+	if err != nil {
+		return nil, err
+	}
+	var rchanges []archive.Change
+	changes, err := c.store.Changes("", layerID)
+	if err == nil {
+		for _, c := range changes {
+			if containerMounts[c.Path] {
+				continue
+			}
+			rchanges = append(rchanges, c)
+		}
+	}
+	return rchanges, err
+}
+
+// To make the checkpoint/restore code use the same fields as Podman:
+type ContainerConfig struct {
+	ID              string `json:"id"`
+	Name            string `json:"name"`
+	RootfsImageName string `json:"rootfsImageName,omitempty"`
+}
+
+// prepareCheckpointExport writes the config and spec to
+// JSON files for later export
+// Podman: libpod/container_internal.go
+func (c *ContainerServer) prepareCheckpointExport(ctr *oci.Container) error {
+	config := &ContainerConfig{
+		ID:              ctr.ID(),
+		Name:            ctr.Name(),
+		RootfsImageName: ctr.ImageName(),
+	}
+
+	if _, err := metadata.WriteJSONFile(config, ctr.Dir(), metadata.ConfigDumpFile); err != nil {
+		return err
+	}
+
+	// save spec
+	jsonPath := filepath.Join(ctr.BundlePath(), "config.json")
+	g, err := generate.NewFromFile(jsonPath)
+	if err != nil {
+		return errors.Wrapf(err, "generating spec for container %q failed", ctr.ID())
+	}
+	if _, err := metadata.WriteJSONFile(g.Config, ctr.Dir(), metadata.SpecDumpFile); err != nil {
+		return errors.Wrapf(err, "generating spec for container %q failed", ctr.ID())
+	}
+
+	return nil
+}
+
+func (c *ContainerServer) exportCheckpoint(ctr *oci.Container, specgen *rspec.Spec, export string) error {
+	id := ctr.ID()
+	dest := ctr.Dir()
+	logrus.Debugf("Exporting checkpoint image of container %q to %q", id, dest)
+
+	includeFiles := []string{
+		"dump.log",
+		metadata.CheckpointDirectory,
+		metadata.ConfigDumpFile,
+		metadata.SpecDumpFile,
+	}
+
+	// To correctly track deleted files, let's go through the output of 'podman diff'
+	rootFsChanges, err := c.getDiff(id)
+	if err != nil {
+		return errors.Wrapf(err, "error exporting root file-system diff for %q", id)
+	}
+	mountPoint, err := c.StorageImageServer().GetStore().Mount(id, specgen.Linux.MountLabel)
+	if err != nil {
+		return errors.Wrapf(err, "Not able to get mountpoint for container %q", id)
+	}
+	addToTarFiles, err := crutils.CRCreateRootFsDiffTar(&rootFsChanges, mountPoint, dest)
+	if err != nil {
+		return err
+	}
+
+	// Put log file into checkpoint archive
+	_, err = os.Stat(specgen.Annotations[annotations.LogPath])
+	if err == nil {
+		src, err := os.Open(specgen.Annotations[annotations.LogPath])
+		if err != nil {
+			return errors.Wrapf(err, "error opening log file %q", specgen.Annotations[annotations.LogPath])
+		}
+		defer src.Close()
+		destLog, err := os.Create(filepath.Join(dest, annotations.LogPath))
```

Comment:
nit: create a new var for `filepath.Join(dest, annotations.LogPath)` and reuse it in the paths below